### PR TITLE
feat: Add ability to disable Tensorboard logging for performance gains [MLG-22]

### PR DESCRIPTION
**Pull request by szewaiyuen6**
## Description

[MLG-22](https://determinedai.atlassian.net/browse/MLG-22)

### Motivation
Tensorboard logging comes with a performance cost of ~10% time per step, this PR add a new feature to enable users to disable automatic Tensorboard logging.

### This PR
This PR composes of the following pieces of work:

(1) Update the experiment config schema with new field tensorboard_logging (default to true to preserve existing behaviors). Run make all to re-generate codes that depend on the config
   - schemas/expconf/v0/experiment.json
   - harness/determined/_info.py
   - harness/determined/common/schemas/expconf/_v0.py
   - harness/determined/_experiment_config.py
   - Generated
       - master/pkg/schemas/expconf/experiment_config.go
       - master/pkg/schemas/expconf/zgen_experiment_config_v0.go
       - master/pkg/schemas/zgen_schemas.go
       - harness/determined/common/schemas/expconf/_gen.py
 
(2) Change core API code path
   - harness/determined/core/_context.py
 
(3) Change trial code path
   - harness/determined/_trial_controller.py
   - harness/determined/estimator/_estimator_trial.py
   - harness/determined/keras/_tf_keras_trial.py
   - harness/determined/pytorch/_pytorch_trial.py
   - harness/determined/pytorch/deepspeed/_deepspeed_trial.py
 
(4) Update tests as needed

## Design Decision
(Feel free to comment on the following points)
- The flag is set on experiment config level.
    - the reason behind this decision is to align with flags of similar nature (e.g. debug). Also, users can easily toggle Tensorboard logging without changing training code in Python
- The flag is a top level argument (as opposed to inside of a struct). I think logically, it would be great to either group this flag with other Tensorboard fields (e.g. Tensorboard Storage) or profiling fields (e.g. profiling). However, regrouping would mean breaking existing agreement on the experiment config, therefore, I am just putting this flag on the top level but open to suggestions.

## local e2e Tests
Run the following test cases with two variation
- Case A: set tensorboarding_logging = false => expect no tensorboard files created
- Case B: not set tensorboard_logging (default behavior) => expect tensorboard files created

### Core API
Use [coreapi_mnist_tutorial_metrics example](https://github.com/determined-ai/determined/tree/master/examples/tutorials/core_api_pytorch_mnist)

Spin up cluster and run locally

**Case A**:
- Verified that no tensorboard files are created
- Verified that trial otherwise completed successfully

**Case B**:
- verified that  tensorboard files are created as before
- Verified that trial otherwise completed successfully

### PyTorch Trial
Use [mnist_pytorch](https://github.com/determined-ai/determined/tree/master/examples/tutorials/mnist_pytorch) (const.yaml) for testing

**Case A**:
- Verified that no tensorboard files are created
- Verified that trial otherwise completed successfully

**Case B**:
- verified that  tensorboard files are created as before
- Verified that trial otherwise completed successfully

### DeepSpeed Trial [To-do]
Use [CIFAR10 MOE example](https://github.com/determined-ai/determined/tree/master/examples/deepspeed/cifar10_moe)
 [Note: running into some issue with deepspeed trial, even on master branch, will test once resolved]

### Keras Trial
Use Keras example [here](https://github.com/determined-ai/determined/tree/master/examples/tutorials/fashion_mnist_tf_keras) 

Spin up cluster and run locally

**Case A**:
- Verified that no tensorboard files are created
- Verified that trial otherwise completed successfully

**Case B**:
- verified that  tensorboard files are created as before
- Verified that trial otherwise completed successfully

### Estimator Trial
Use [gbt titanic estimator](https://github.com/determined-ai/determined/blob/master/examples/decision_trees/gbt_titanic_estimator/model_def.py)

**Case A**:
- Verified that only two tensorboard files are created (EstimatorTrial has other paths of adding TF files at the end of experiment, we are only disabling the file upload during training, which cuts down TB files upload to two and serves the purpose of performance enhancement)
- Verified that trial otherwise completed successfully

**Case B**:
- verified that  tensorboard files are created as before
- Verified that trial otherwise completed successfully

## Distributed e2e test 
Deploy `c36f2260c82b14a2f9c702d7b1cbff794644601b` to AWS
Run distributed.yaml from [PyTorch MNIST example](https://github.com/determined-ai/determined/tree/master/examples/tutorials/mnist_pytorch)

**Case A**:
- Verified that no tensorboard files are created
- Verified that trial otherwise completed successfully

**Case B**:
- verified that  tensorboard files are created as before
- Verified that trial otherwise completed successfully

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
MLG-22

[MLG-22]: https://determinedai.atlassian.net/browse/MLG-22?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by szewaiyuen6** [comment]
Closing PR as this has dependency on trainer API and BatchMetricWriter refactoring. After both changes, the implementation here will be different.

Will create new PR later once dependency is finished.
